README.md
# Dekajs

> DEKAtrian for JavaScript

This is a simple library for converting dates between [Dekatrian](https://www.facebook.com/dekatrian/) and [Gregorian](https://en.wikipedia.org/wiki/Adoption_of_the_Gregorian_calendar) calendars. This library is not official.

This package is compatible with [StandardJS](https://standardjs.com) and [LibreJS](https://www.gnu.org/software/librejs).

## Install

This project uses [Node.js](https://nodejs.org) and [npm](https://www.npmjs.com).

``` bash
$ npm install @sepbit/dekajs
```

## Usage

CommonJS module.

``` javascript
const dekatrian = require('@sepbit/dekajs')

// Convert from Dekatrian to Gregorian calendars
dekatrian.dekaToGreg(2018, 13, 28)
// Return string, 2018-12-31

// Convert from Gregorian to Dekatrian calendars
dekatrian.gregToDeka(2018, 12, 31)
// Return string, 2018-13-28
```

ECMAScript module.


``` html
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Title</title>
  </head>
  <body>
    <script type="module">
      import { dekaToGreg, gregToDeka } from 'https://unpkg.com/@sepbit/dekajs/src/main.js'

      // Convert from Dekatrian to Gregorian calendars
      dekaToGreg(2018, 13, 28)
      // Return string, 2018-12-31

      // Convert from Gregorian to Dekatrian calendars
      gregToDeka(2018, 12, 31)
      // Return string, 2018-13-28
    </script>
  </body>
</html>
```

## Contributing

Pull Requests not accepted.

## Security

If you discover any security related issues, please email `contato@sepbit.com` instead of using the issue tracker.

## License

GPL-3.0-or-later, please see [COPYING](COPYING) file for more information.


CHANGELOG.md
# Changelog

See [Keep a Changelog](http://keepachangelog.com/).

## [8.0.0] 2019-12-27
### Changed
- Change directory structure

## [7.1.0] 2019-10-12
### Changed
- Change from NotABug to GitLab

## [7.0.2] 2019-09-25
### Changed
- Minor improvements

## [7.0.1] 2019-08-18
### Changed
- notice license

## [7.0.0] 2019-08-16
### Changed
- New release

## [6.0.4] 2019-05-24
### Changed
- Change from GitHub to NotABug

## [6.0.3] 2019-03-16
### Changed
- Information for contribution

### Removed
- CODE_OF_CONDUCT.md CONTRIBUTING.md files

## [6.0.2] 2019-03-06
### Fixed
- README.md, example CJS

## [6.0.1] 2019-03-06
### Fixed
- Coverage

## [6.0.0] 2019-03-05
### Changed
- export default to export

## [5.0.8] 2019-02-16
### Changed
- Update package.json
- dekatrian_test.mjs to dekatrianTest.mjs

## [5.0.7] 2019-02-16
### Changed
- Update README.md

## [5.0.6] 2019-02-16
### Changed
- Update CONTRIBUTING.md

## [5.0.5] 2019-02-16
### Changed
- Scrutinizer CI to Travis CI and Coveralls

## [5.0.4] 2019-02-12
### Fixed
- Standard.JS style guide

## [5.0.3] 2019-02-12
### Changed
- Appropriate legal notice

## [5.0.2] - 2019-02-11
### Changed
- test/*.js to test/*.mjs

## [5.0.1] - 2019-02-10
### Fixed
- dekaToGreg() function, error Sinchronian day

## [5.0.0] - 2019-01-29
### Added
- ECMAScript module, with [ESM](https://github.com/standard-things/esm)

### Changed
- testing framework [MOCHA](https://github.com/mochajs/mocha) to [AVA](https://github.com/avajs/ava)

## [4.0.7] - 2018-12-24
### Changed
- engine [NPM](https://github.com/npm/cli) ^6.4 to ^5.8

### Removed
- tags file

## [4.0.6] - 2018-12-22
### Changed
- assert.equal() to assert.strictEqual()

## [4.0.5] - 2018-12-22
### Changed
- Appropriate legal notices

## [4.0.4] 2018-10-29
### Changed
- Use JSDoc

## [4.0.3] 2018-10-20
### Changed
- Universal Ctags

### Removed
- Code Intelligence badge in README.md

## [4.0.2] 2018-10-20
### Changed
- indent style

## [4.0.1] 2018-09-14
### Fixed
- Get Gregorian day of year

## [4.0.0] 2018-09-13
### Changed
- dekatoGreg method to dekaToGreg

## [3.0.1] - 2018-06-20
### Changed
- Publish the readme file in NPM.js

## [3.0.0] - 2018-06-17
### Changed
- changes for Scoped Packages

## [2.0.0] - 2018-04-08
### Changed
- Change to static methods

## [1.0.1] - 2018-02-11
### Changed
- package.json

## [1.0.0] - 2018-02-11
### Added
- First release!


test/dekatrianTest.js
/**
 * @licstart  The following is the entire license notice for the
 * JavaScript code in this page.
 *
 * Dekajs - DEKAtrian for JavaScript
 * Copyright (C) 2017-2019  Sepbit
 *
 * The JavaScript code in this page is free software: you can
 * redistribute it and/or modify it under the terms of the GNU
 * General Public License (GNU GPL) as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option)
 * any later version.  The code is distributed WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
 *
 * As additional permission under GNU GPL version 3 section 7, you
 * may distribute non-source (e.g., minimized or compacted) forms of
 * that code without the copy of the GNU GPL normally required by
 * section 4, provided you include this license notice and a URL
 * through which recipients can access the Corresponding Source.
 *
 * @licend  The above is the entire license notice
 * for the JavaScript code in this page.
 */

import test from 'ava'
import { dekaToGreg, gregToDeka } from '../src/index.js'

/** Test private function checkLeapYear */
test('400 year', t => {
  t.is(gregToDeka(400, 1, 1), '400-0-1')
})

/** Test private function checkGregorian */
test('Error 2016-2-30', t => {
  const error = t.throws(() => {
    gregToDeka(2016, 2, 30)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

test('Error 2018-2-29', t => {
  const error = t.throws(() => {
    gregToDeka(2018, 2, 29)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

test('Error 2018-1-32', t => {
  const error = t.throws(() => {
    gregToDeka(2018, 1, 32)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

test('Error 2018-13-31', t => {
  const error = t.throws(() => {
    gregToDeka(2018, 4, 31)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

/** Test private function checkDekatrian */
test('Error 2016-0-3', t => {
  const error = t.throws(() => {
    dekaToGreg(2016, 0, 3)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Dekatrian date')
})

test('Error 2018-0-2', t => {
  const error = t.throws(() => {
    dekaToGreg(2018, 0, 2)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Dekatrian date')
})

test('Error 2018-14-2', t => {
  const error = t.throws(() => {
    dekaToGreg(2018, 14, 2)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Dekatrian date')
})

/** Test public function dekaToGreg */
test('dekaToGreg - Achronian day', t => {
  t.is(dekaToGreg(2018, 0, 1), '2018-1-1')
})

test('dekaToGreg - Sinchronian day', t => {
  t.is(dekaToGreg(2016, 0, 2), '2016-1-2')
})

test('dekaToGreg - Other day', t => {
  t.is(dekaToGreg(2018, 13, 28), '2018-12-31')
})

/** Test public function gregToDeka */
test('gregToDeka - Achronian day', t => {
  t.is(gregToDeka(2018, 1, 1), '2018-0-1')
})

test('gregToDeka - Sinchronian day', t => {
  t.is(gregToDeka(2016, 1, 2), '2016-0-2')
})

test('gregToDeka - Other day', t => {
  t.is(gregToDeka(2018, 12, 31), '2018-13-28')
})


test/dekatrianTest.mjs
/**
 * @source: https://github.com/vitorteccom/dekajs
 *
 * @licstart  The following is the entire license notice for the
 * JavaScript code in this page.
 *
 * Dekajs - DEKAtrian for JavaScript
 * Copyright (C) 2017-2019  Vitortec.com
 *
 * The JavaScript code in this page is free software: you can
 * redistribute it and/or modify it under the terms of the GNU
 * General Public License (GNU GPL) as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option)
 * any later version.  The code is distributed WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
 *
 * As additional permission under GNU GPL version 3 section 7, you
 * may distribute non-source (e.g., minimized or compacted) forms of
 * that code without the copy of the GNU GPL normally required by
 * section 4, provided you include this license notice and a URL
 * through which recipients can access the Corresponding Source.
 *
 * @licend  The above is the entire license notice
 * for the JavaScript code in this page.
 *
 * ECMAScript version 6
 *
 * @author    Vitor Guia <vitor.guia@vitortec.com>
 * @copyright 2017-2019 Vitortec.com
 * @license   GPL-3.0-or-later
 * @see       {@link https://github.com/vitorteccom/dekajs|Repository of Dekajs}
 */

import test from 'ava'
import { dekaToGreg, gregToDeka } from '../lib/main.mjs'

/** Test private function checkLeapYear */
test('400 year', t => {
  t.is(gregToDeka(400, 1, 1), '400-0-1')
})

/** Test private function checkGregorian */
test('Error 2016-2-30', t => {
  const error = t.throws(() => {
    gregToDeka(2016, 2, 30)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

test('Error 2018-2-29', t => {
  const error = t.throws(() => {
    gregToDeka(2018, 2, 29)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

test('Error 2018-1-32', t => {
  const error = t.throws(() => {
    gregToDeka(2018, 1, 32)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

test('Error 2018-13-31', t => {
  const error = t.throws(() => {
    gregToDeka(2018, 4, 31)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

/** Test private function checkDekatrian */
test('Error 2016-0-3', t => {
  const error = t.throws(() => {
    dekaToGreg(2016, 0, 3)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Dekatrian date')
})

test('Error 2018-0-2', t => {
  const error = t.throws(() => {
    dekaToGreg(2018, 0, 2)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Dekatrian date')
})

test('Error 2018-14-2', t => {
  const error = t.throws(() => {
    dekaToGreg(2018, 14, 2)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Dekatrian date')
})

/** Test public function dekaToGreg */
test('dekaToGreg - Achronian day', t => {
  t.is(dekaToGreg(2018, 0, 1), '2018-1-1')
})

test('dekaToGreg - Sinchronian day', t => {
  t.is(dekaToGreg(2016, 0, 2), '2016-1-2')
})

test('dekaToGreg - Other day', t => {
  t.is(dekaToGreg(2018, 13, 28), '2018-12-31')
})

/** Test public function gregToDeka */
test('gregToDeka - Achronian day', t => {
  t.is(gregToDeka(2018, 1, 1), '2018-0-1')
})

test('gregToDeka - Sinchronian day', t => {
  t.is(gregToDeka(2016, 1, 2), '2016-0-2')
})

test('gregToDeka - Other day', t => {
  t.is(gregToDeka(2018, 12, 31), '2018-13-28')
})


test/dekatrian_test.mjs
/**
 * @source: https://github.com/vitorteccom/dekajs
 *
 * @licstart  The following is the entire license notice for the
 * JavaScript code in this page.
 *
 * Dekajs - DEKAtrian for JavaScript
 * Copyright (C) 2017-2019  Vitortec.com
 *
 * The JavaScript code in this page is free software: you can
 * redistribute it and/or modify it under the terms of the GNU
 * General Public License (GNU GPL) as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option)
 * any later version.  The code is distributed WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
 *
 * As additional permission under GNU GPL version 3 section 7, you
 * may distribute non-source (e.g., minimized or compacted) forms of
 * that code without the copy of the GNU GPL normally required by
 * section 4, provided you include this license notice and a URL
 * through which recipients can access the Corresponding Source.
 *
 * @licend  The above is the entire license notice
 * for the JavaScript code in this page.
 *
 * ECMAScript version 6
 *
 * @author    Vitor Guia <vitor.guia@vitortec.com>
 * @copyright 2017-2019 Vitortec.com
 * @license   GPL-3.0-or-later
 * @see       {@link https://github.com/vitorteccom/dekajs|Repository of Dekajs}
 */

import test from 'ava'
import dekatrian from '../lib/main'

/** Test private function checkLeapYear */
test('400 year', t => {
  t.is(dekatrian.gregToDeka(400, 1, 1), '400-0-1')
})

/** Test private function checkGregorian */
test('Error 2016-2-30', t => {
  const error = t.throws(() => {
    dekatrian.gregToDeka(2016, 2, 30)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

test('Error 2018-2-29', t => {
  const error = t.throws(() => {
    dekatrian.gregToDeka(2018, 2, 29)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

test('Error 2018-1-32', t => {
  const error = t.throws(() => {
    dekatrian.gregToDeka(2018, 1, 32)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

test('Error 2018-13-31', t => {
  const error = t.throws(() => {
    dekatrian.gregToDeka(2018, 4, 31)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Gregorian date')
})

/** Test private function checkDekatrian */
test('Error 2016-0-3', t => {
  const error = t.throws(() => {
    dekatrian.dekaToGreg(2016, 0, 3)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Dekatrian date')
})

test('Error 2018-0-2', t => {
  const error = t.throws(() => {
    dekatrian.dekaToGreg(2018, 0, 2)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Dekatrian date')
})

test('Error 2018-14-2', t => {
  const error = t.throws(() => {
    dekatrian.dekaToGreg(2018, 14, 2)
  }, Error)

  t.is(error.message, 'The provider arguments are not a valid Dekatrian date')
})

/** Test public function dekaToGreg */
test('dekaToGreg - Achronian day', t => {
  t.is(dekatrian.dekaToGreg(2018, 0, 1), '2018-1-1')
})

test('dekaToGreg - Sinchronian day', t => {
  t.is(dekatrian.dekaToGreg(2016, 0, 2), '2016-1-2')
})

test('dekaToGreg - Other day', t => {
  t.is(dekatrian.dekaToGreg(2018, 13, 28), '2018-12-31')
})

/** Test public function gregToDeka */
test('gregToDeka - Achronian day', t => {
  t.is(dekatrian.gregToDeka(2018, 1, 1), '2018-0-1')
})

test('gregToDeka - Sinchronian day', t => {
  t.is(dekatrian.gregToDeka(2016, 1, 2), '2016-0-2')
})

test('gregToDeka - Other day', t => {
  t.is(dekatrian.gregToDeka(2018, 12, 31), '2018-13-28')
})


lib/main.js
/**
 * @licstart  The following is the entire license notice for the
 * JavaScript code in this page.
 *
 * Dekajs - DEKAtrian for JavaScript
 * Copyright (C) 2017-2019  Sepbit
 *
 * The JavaScript code in this page is free software: you can
 * redistribute it and/or modify it under the terms of the GNU
 * General Public License (GNU GPL) as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option)
 * any later version.  The code is distributed WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
 *
 * As additional permission under GNU GPL version 3 section 7, you
 * may distribute non-source (e.g., minimized or compacted) forms of
 * that code without the copy of the GNU GPL normally required by
 * section 4, provided you include this license notice and a URL
 * through which recipients can access the Corresponding Source.
 *
 * @licend  The above is the entire license notice
 * for the JavaScript code in this page.
 *
 * @access private
 */

import { dekaToGreg, gregToDeka } from './dekatrian.js'
export { dekaToGreg, gregToDeka }


lib/index.js
/* global require:writable */
/**
 * @licstart  The following is the entire license notice for the
 * JavaScript code in this page.
 *
 * Dekajs - DEKAtrian for JavaScript
 * Copyright (C) 2017-2019  Sepbit
 *
 * The JavaScript code in this page is free software: you can
 * redistribute it and/or modify it under the terms of the GNU
 * General Public License (GNU GPL) as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option)
 * any later version.  The code is distributed WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
 *
 * As additional permission under GNU GPL version 3 section 7, you
 * may distribute non-source (e.g., minimized or compacted) forms of
 * that code without the copy of the GNU GPL normally required by
 * section 4, provided you include this license notice and a URL
 * through which recipients can access the Corresponding Source.
 *
 * @licend  The above is the entire license notice
 * for the JavaScript code in this page.
 *
 * @access private
 */

require = require('esm')(module/*, options */)
module.exports = require('./main.js')


lib/main.mjs
/**
 * @source: https://notabug.org/sepbit/dekajs
 *
 * @licstart  The following is the entire license notice for the
 * JavaScript code in this page.
 *
 * Dekajs - DEKAtrian for JavaScript
 * Copyright (C) 2017-2019  SEPBIT
 *
 * The JavaScript code in this page is free software: you can
 * redistribute it and/or modify it under the terms of the GNU
 * General Public License (GNU GPL) as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option)
 * any later version.  The code is distributed WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
 *
 * As additional permission under GNU GPL version 3 section 7, you
 * may distribute non-source (e.g., minimized or compacted) forms of
 * that code without the copy of the GNU GPL normally required by
 * section 4, provided you include this license notice and a URL
 * through which recipients can access the Corresponding Source.
 *
 * @licend  The above is the entire license notice
 * for the JavaScript code in this page.
 *
 * ECMAScript version 6
 *
 * @author    Vitor Guia <contato@vitor.guia.nom.br>
 * @copyright 2017-2019 SEPBIT
 * @license   GPL-3.0-or-later
 * @see       {@link https://notabug.org/sepbit/dekajs|Repository of Dekajs}
 * @access    private
 */

import { dekaToGreg, gregToDeka } from './dekatrian.mjs'
export { dekaToGreg, gregToDeka }


src/dekatrian.js
/**
 * @licstart  The following is the entire license notice for the
 * JavaScript code in this page.
 *
 * Dekajs - DEKAtrian for JavaScript
 * Copyright (C) 2017-2019  Sepbit
 *
 * The JavaScript code in this page is free software: you can
 * redistribute it and/or modify it under the terms of the GNU
 * General Public License (GNU GPL) as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option)
 * any later version.  The code is distributed WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
 *
 * As additional permission under GNU GPL version 3 section 7, you
 * may distribute non-source (e.g., minimized or compacted) forms of
 * that code without the copy of the GNU GPL normally required by
 * section 4, provided you include this license notice and a URL
 * through which recipients can access the Corresponding Source.
 *
 * @licend  The above is the entire license notice
 * for the JavaScript code in this page.
 *
 * @module Dekatrian
 */

/**
 * Check if a year is leap year or not
 *
 * @access private
 *
 * @param {integer} year Year
 *
 * @return {boolean} Year is leap year or not
 */
function checkLeapYear (year) {
  if (year % 4 === 0 && year % 100 !== 0) {
    return true
  }
  if (year % 400 === 0) {
    return true
  }
  return false
}

/**
 * Get Gregorian day of year
 *
 * @access private
 *
 * @param {integer} year  Gregorian year
 * @param {integer} month Gregorian month
 * @param {integer} day   Gregorian day
 *
 * @return {integer} Gregorian day of year
 */
function getGregorianYearDay (year, month, day) {
  month--
  let end = new Date()
  end = end.setFullYear(year, month, day)
  let start = new Date()
  start = start.setFullYear(year, 0, 0)
  return Math.round((end - start) / 864e5)
}

/**
 * Check if a Gregorian date is valid or not
 *
 * @access private
 *
 * @param {integer} year  Gregorian year
 * @param {integer} month Gregorian month
 * @param {integer} day   Gregorian day
 *
 * @return {boolean} Gregorian date is valid or not
 */
function checkGregorian (year, month, day) {
  if (month === 2 && checkLeapYear(year) && day > 29) {
    return false
  }
  if (month === 2 && !checkLeapYear(year) && day > 28) {
    return false
  }
  if (month === 1 || month === 3 || month === 5 || month === 7 ||
      month === 8 || month === 10 || month === 12) {
    if (day > 31) {
      return false
    }
  } else {
    if (month > 12 || day > 30) {
      return false
    }
  }
  return true
}

/**
 * Check if a Dekatrian date is valid or not
 *
 * @access private
 *
 * @param {integer} year  Dekatrian year
 * @param {integer} month Dekatrian month
 * @param {integer} day   Dekatrian day
 *
 * @return {boolean} Dekatrian date is valid or not
 */
function checkDekatrian (year, month, day) {
  if (checkLeapYear(year)) {
    if (month === 0 && day > 2) {
      return false
    }
  } else {
    if (month === 0 && day > 1) {
      return false
    }
  }
  if (month > 13 || day > 28) {
    return false
  }
  return true
}

/**
 * Convert a Dekatrian date to Gregorian date
 *
 * @param {integer} year  Dekatrian year
 * @param {integer} month Dekatrian month
 * @param {integer} day   Dekatrian day
 *
 * @throws {Error} Is not a valid Dekatrian date
 *
 * @return {string} Gregorian date e.g. 2018-12-31
 */
function dekaToGreg (year, month, day) {
  if (!checkDekatrian(year, month, day)) {
    throw new Error('The provider arguments are not a valid Dekatrian date')
  }

  /** Achronian day */
  if (month === 0 && day === 1) {
    return year + '-1-1'
  }

  /** leap year */
  if (checkLeapYear(year)) {
    /** Sinchronian day */
    if (month === 0 && day === 2) {
      return year + '-1-2'
    }

    day++
  }

  const dayYear = (month * 28) - 28 + day + 1
  let gregDate = new Date(year, 0)
  gregDate = new Date(gregDate.setDate(dayYear))
  const gregYear = gregDate.getFullYear()
  const gregMonth = gregDate.getMonth() + 1
  const gregDay = gregDate.getDate()

  return gregYear + '-' + gregMonth + '-' + gregDay
}

/**
 * Convert a Gregorian date to Dekatrian date
 *
 * @param {integer} year  Gregorian year
 * @param {integer} month Gregorian month
 * @param {integer} day   Gregorian day
 *
 * @throws {Error} Is not a valid Gregorian date
 *
 * @return {string} Dekatrian date e.g. 2018-13-28
 */
function gregToDeka (year, month, day) {
  if (!checkGregorian(year, month, day)) {
    throw new Error('The provider arguments are not a valid Gregorian date')
  }

  let yearDay = getGregorianYearDay(year, month, day) - 1

  /** leap year */
  if (checkLeapYear(year)) {
    yearDay--

    /** Sinchronian day */
    if (month === 1 && day === 2) {
      return year + '-0-2'
    }
  }

  /** Achronian day */
  if (month === 1 && day === 1) {
    return year + '-0-1'
  }

  let dekaMonth = Math.floor(yearDay / 28)
  let dekaDay = yearDay - (dekaMonth * 28)
  dekaMonth++

  if (!dekaDay) {
    dekaDay = 28
    dekaMonth--
  }

  return year + '-' + dekaMonth + '-' + dekaDay
}

/** ECMAScript module */
export { dekaToGreg, gregToDeka }


lib/dekatrian.mjs
/**
 * @source: https://notabug.org/sepbit/dekajs
 *
 * @licstart  The following is the entire license notice for the
 * JavaScript code in this page.
 *
 * Dekajs - DEKAtrian for JavaScript
 * Copyright (C) 2017-2019  SEPBIT
 *
 * The JavaScript code in this page is free software: you can
 * redistribute it and/or modify it under the terms of the GNU
 * General Public License (GNU GPL) as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option)
 * any later version.  The code is distributed WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
 *
 * As additional permission under GNU GPL version 3 section 7, you
 * may distribute non-source (e.g., minimized or compacted) forms of
 * that code without the copy of the GNU GPL normally required by
 * section 4, provided you include this license notice and a URL
 * through which recipients can access the Corresponding Source.
 *
 * @licend  The above is the entire license notice
 * for the JavaScript code in this page.
 *
 * ECMAScript version 6
 *
 * @author    Vitor Guia <contato@vitor.guia.nom.br>
 * @copyright 2017-2019 SEPBIT
 * @license   GPL-3.0-or-later
 * @see       {@link https://notabug.org/sepbit/dekajs|Repository of Dekajs}
 * @module    Dekatrian
 */

/**
 * Check if a year is leap year or not
 *
 * @access private
 *
 * @param {integer} year Year
 *
 * @return {boolean} Year is leap year or not
 */
function checkLeapYear (year) {
  if (year % 4 === 0 && year % 100 !== 0) {
    return true
  }
  if (year % 400 === 0) {
    return true
  }
  return false
}

/**
 * Get Gregorian day of year
 *
 * @access private
 *
 * @param {integer} year  Gregorian year
 * @param {integer} month Gregorian month
 * @param {integer} day   Gregorian day
 *
 * @return {integer} Gregorian day of year
 */
function getGregorianYearDay (year, month, day) {
  month--
  let end = new Date()
  end = end.setFullYear(year, month, day)
  let start = new Date()
  start = start.setFullYear(year, 0, 0)
  return Math.round((end - start) / 864e5)
}

/**
 * Check if a Gregorian date is valid or not
 *
 * @access private
 *
 * @param {integer} year  Gregorian year
 * @param {integer} month Gregorian month
 * @param {integer} day   Gregorian day
 *
 * @return {boolean} Gregorian date is valid or not
 */
function checkGregorian (year, month, day) {
  if (month === 2 && checkLeapYear(year) && day > 29) {
    return false
  }
  if (month === 2 && !checkLeapYear(year) && day > 28) {
    return false
  }
  if (month === 1 || month === 3 || month === 5 || month === 7 ||
      month === 8 || month === 10 || month === 12) {
    if (day > 31) {
      return false
    }
  } else {
    if (month > 12 || day > 30) {
      return false
    }
  }
  return true
}

/**
 * Check if a Dekatrian date is valid or not
 *
 * @access private
 *
 * @param {integer} year  Dekatrian year
 * @param {integer} month Dekatrian month
 * @param {integer} day   Dekatrian day
 *
 * @return {boolean} Dekatrian date is valid or not
 */
function checkDekatrian (year, month, day) {
  if (checkLeapYear(year)) {
    if (month === 0 && day > 2) {
      return false
    }
  } else {
    if (month === 0 && day > 1) {
      return false
    }
  }
  if (month > 13 || day > 28) {
    return false
  }
  return true
}

/**
 * Convert a Dekatrian date to Gregorian date
 *
 * @param {integer} year  Dekatrian year
 * @param {integer} month Dekatrian month
 * @param {integer} day   Dekatrian day
 *
 * @throws {Error} If not a valid Dekatrian date
 *
 * @return {string} Gregorian date e.g. 2018-12-31
 */
function dekaToGreg (year, month, day) {
  if (!checkDekatrian(year, month, day)) {
    throw new Error('The provider arguments are not a valid Dekatrian date')
  }

  /** Achronian day */
  if (month === 0 && day === 1) {
    return year + '-1-1'
  }

  /** leap year */
  if (checkLeapYear(year)) {
    /** Sinchronian day */
    if (month === 0 && day === 2) {
      return year + '-1-2'
    }

    day++
  }

  let dayYear = (month * 28) - 28 + day + 1
  let gregDate = new Date(year, 0)
  gregDate = new Date(gregDate.setDate(dayYear))
  let gregYear = gregDate.getFullYear()
  let gregMonth = gregDate.getMonth() + 1
  let gregDay = gregDate.getDate()

  return gregYear + '-' + gregMonth + '-' + gregDay
}

/**
 * Convert a Gregorian date to Dekatrian date
 *
 * @param {integer} year  Gregorian year
 * @param {integer} month Gregorian month
 * @param {integer} day   Gregorian day
 *
 * @throws {Error} If not a valid Gregorian date
 *
 * @return {string} Dekatrian date e.g. 2018-13-28
 */
function gregToDeka (year, month, day) {
  if (!checkGregorian(year, month, day)) {
    throw new Error('The provider arguments are not a valid Gregorian date')
  }

  let yearDay = getGregorianYearDay(year, month, day) - 1

  /** leap year */
  if (checkLeapYear(year)) {
    yearDay--

    /** Sinchronian day */
    if (month === 1 && day === 2) {
      return year + '-0-2'
    }
  }

  /** Achronian day */
  if (month === 1 && day === 1) {
    return year + '-0-1'
  }

  let dekaMonth = Math.floor(yearDay / 28)
  let dekaDay = yearDay - (dekaMonth * 28)
  dekaMonth++

  if (!dekaDay) {
    dekaDay = 28
    dekaMonth--
  }

  return year + '-' + dekaMonth + '-' + dekaDay
}

/** ECMAScript module */
export { dekaToGreg, gregToDeka }
